coreax/kernels/base.py

```diff
@@ -0,0 +1,503 @@
+
+r"""
+Classes and associated functionality to use kernel functions.
+
+A kernel is a non-negative, real-valued integrable function that can take two inputs,
+``x`` and ``y``, and returns a value that decreases as ``x`` and ``y`` move further away
+in space from each other. Note that *further* here may account for cyclic behaviour in
+the data, for example.
+
+In this library, we often use kernels as a smoothing tool: given a dataset of distinct
+points, we can reconstruct the underlying data generating distribution through smoothing
+of the data with kernels.
+
+Some kernels are parameterizable and may represent other well known kernels when given
+appropriate parameter values. For example, the
+:class:`~coreax.kernels.SquaredExponentialKernel`,
+
+.. math::
+
+    k(x,y) = \text{output_scale} * \exp (-||x-y||^2/2 * \text{length_scale}^2),
+
+which if parameterized with an ``output_scale`` of
+:math:`\frac{1}{\sqrt{2\pi} \,*\, \text{length_scale}}`, yields the well known
+Gaussian kernel.
+
+A :class:`~coreax.kernels.ScalarValuedKernel` must implement a
+:meth:`~coreax.kernels.ScalarValuedKernel.compute_elementwise` method, which returns the
+floating point value after evaluating the kernel on two floats, ``x`` and ``y``.
+Additional methods, such as :meth:`ScalarValuedKernel.grad_x_elementwise`, can
+optionally be overridden to improve performance. The canonical example is when a
+suitable closed-form representation of a higher-order gradient can be used to avoid the
+expense of performing automatic differentiation.
+
+Such an example can be seen in :class:`coreax.kernels.SteinKernel`, where the analytic
+forms of :meth:`ScalarValuedKernel.divergence_x_grad_y` are significantly cheaper to
+compute than the automatic differentiated default.
+"""
+
+# Support class typing annotations inside itself
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import Union
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+from jax import Array, grad, jacrev
+from jax.typing import ArrayLike
+from typing_extensions import override
+
+from coreax.data import Data, is_data
+from coreax.kernels.util import block_data_convert
+from coreax.util import pairwise, tree_leaves_repeat
+
+
+class ScalarValuedKernel(eqx.Module):
+    """Abstract base class for scalar-valued kernels."""
+
+    def __add__(self, addition: ScalarValuedKernel) -> AdditiveKernel:
+        """Overload `+` operator."""
+        if not isinstance(addition, ScalarValuedKernel):
+            raise ValueError("'addition' must be an instance of a 'ScalarValuedKernel'")
+        return AdditiveKernel(self, addition)
+
+    def __mul__(self, product: ScalarValuedKernel) -> ProductKernel:
+        """Overload `*` operator."""
+        if not isinstance(product, ScalarValuedKernel):
+            raise ValueError("'product' must be an instance of a 'ScalarValuedKernel'")
+        return ProductKernel(self, product)
```

Comment:
We've also lost the ability to directly multiply a kernel by a scalar. (And we'll need to reimplement `__rmul__` to make it work both ways).

Reply:
See above comment

Reply:
Added back in